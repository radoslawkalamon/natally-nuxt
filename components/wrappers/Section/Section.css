.section {
  display: flex;
  flex-flow: column nowrap;
  gap: var(--gap);
}
