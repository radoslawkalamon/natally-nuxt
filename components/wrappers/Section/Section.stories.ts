import type { Meta, Story } from '@storybook/vue'
import WrappersSection from '@/components/wrappers/Section/Section.vue'

const meta: Meta = {
  title: 'Wrappers / Section',
  component: WrappersSection,
  argTypes: {
    header: {
      description: 'hidden when no children',
      table: {
        type: null,
        defaultValue: null
      }
    },
    default: {
      description: '-',
      table: {
        type: null,
        defaultValue: null
      }
    },
    footer: {
      description: 'hidden when no children',
      table: {
        type: null,
        defaultValue: null
      }
    }
  }
}
export default meta

export const allSlots: Story = () => ({
  components: {
    WrappersSection
  },
  template: `<WrappersSection>
    <template #header>
      Header slot
    </template>
    <template>
      Default slot
    </template>
    <template #footer>
      Footer slot
    </template>
  </WrappersSection>`
})

export const noHeader: Story = () => ({
  components: {
    WrappersSection
  },
  template: `<WrappersSection>
    <template>
      Default slot
    </template>
    <template #footer>
      Footer slot
    </template>
  </WrappersSection>`
})

export const noFooter: Story = () => ({
  components: {
    WrappersSection
  },
  template: `<WrappersSection>
    <template #header>
      Header slot
    </template>
    <template>
      Default slot
    </template>
  </WrappersSection>`
})

export const noHeaderNoFooter: Story = () => ({
  components: {
    WrappersSection
  },
  template: `<WrappersSection>
    Default slot
  </WrappersSection>`
})
