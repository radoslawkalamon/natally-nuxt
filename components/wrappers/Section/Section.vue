<template>
  <section class="section">
    <div
      v-if="$slots.header"
      class="section__header"
      data-test="wrappers-section-header"
    >
      <slot name="header" />
    </div>
    <div class="section__inner">
      <slot />
    </div>
    <div
      v-if="$slots.footer"
      class="section__footer"
      data-test="wrappers-section-footer"
    >
      <slot name="footer" />
    </div>
  </section>
</template>

<script src="./Section.ts" lang="ts"></script>
<style src="./Section.css" scoped></style>
