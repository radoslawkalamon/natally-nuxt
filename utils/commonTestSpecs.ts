import Vue from 'vue'
import { mount } from '@vue/test-utils'
import { merge } from 'lodash'

export function shallRender (component: typeof Vue, dataTestID: string, defaultOptions?: object): void {
  test('shall render', (): void => {
    const wrapper = mount(component, defaultOptions)
    const isWrapperVisible = wrapper.find(`[data-test="${dataTestID}"]`).isVisible()
    expect(isWrapperVisible).toBeTruthy()
  })
}

export function shallHaveSlot (component: typeof Vue, slot: string, defaultOptions?: object): void {
  test(`shall have slot '${slot}'`, (): void => {
    const text = `This is slot '${slot}' test.`
    const wrapperOptions = merge(defaultOptions, {
      slots: {
        [slot]: text
      }
    })
    const wrapper = mount(component, wrapperOptions)
    expect(wrapper.html()).toContain(text)
  })
}

export function shallHaveStringProp (component: typeof Vue, propName: string, defaultOptions?: object): void {
  test(`shall have string prop '${propName}'`, (): void => {
    const text = `This is prop ${propName} test.`
    const wrapperOptions = merge(defaultOptions, {
      propsData: {
        [propName]: text
      }
    })
    const wrapper = mount(component, wrapperOptions)
    expect(wrapper.html()).toContain(text)
  })
}

export function shallHavePropInfluenceOverClassNames (
  component: typeof Vue,
  propName: string,
  propValue: unknown,
  classNameSuffix: string,
  defaultOptions?: object
) {
  test(`shall have ${typeof propValue} prop '${propName}' influence over class names`, () => {
    const wrapperOptions = merge(defaultOptions, {
      propsData: {
        [propName]: propValue
      }
    })
    const wrapper = mount(component, wrapperOptions)
    const isWrapperHaveClass = wrapper.classes().some(cls => cls.includes(classNameSuffix))
    expect(isWrapperHaveClass).toBeTruthy()
  })
}
