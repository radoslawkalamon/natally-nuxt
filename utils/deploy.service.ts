/* eslint-disable no-console */
import { Client, FileInfo, FTPResponse } from 'basic-ftp'
import { ProgressInfo } from 'basic-ftp/dist/ProgressTracker'
import capitalize from './formatters/capitalize'
import spacelize from './formatters/spacelize'

type DeployServiceConfig = {
  host: string;
  port: number;
  username: string;
  password: string;
  remoteCatalog: string;
  remoteIgnoreFile: string[];
  localCatalog: string;
}

type RemoveStrategies = {
  [key in FileInfo['type']]: (name: FileInfo['name']) => Promise<void | FTPResponse>
}

function decoratorLogCall () {
  return function (_target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    const targetMethod = descriptor.value

    descriptor.value = function (...args: any[]) {
      const formattedName = spacelize(capitalize(propertyKey))
      console.log(`::: ${formattedName} :::`)
      return targetMethod.apply(this, args)
    }

    return descriptor
  }
}

class DeployService {
  readonly host: string;
  readonly port: number;
  readonly username: string;
  readonly password: string;
  readonly remoteCatalog: string;
  readonly remoteIgnoreFile: string[];
  readonly localCatalog: string;
  FTPClient = new Client()

  constructor (config: DeployServiceConfig) {
    this.host = config.host
    this.port = Number(config.port)
    this.username = config.username
    this.password = config.password
    this.remoteCatalog = config.remoteCatalog
    this.remoteIgnoreFile = config.remoteIgnoreFile
    this.localCatalog = config.localCatalog
  }

  public async start (): Promise<void> {
    try {
      await this.openConnection()
      await this.ensureRemoteCatalog()
      await this.openRemoteCatalog()
      await this.cleanRemoteCatalog()
      await this.uploadToRemoteCatalog()
    } catch (e) {
      console.error(e)
      throw e
    } finally {
      this.closeConnection()
    }
  }

  @decoratorLogCall()
  private async openConnection (): Promise<void> {
    await this.FTPClient.access({
      host: this.host,
      port: this.port,
      user: this.username,
      password: this.password,
      secure: false
    })
  }

  @decoratorLogCall()
  private async ensureRemoteCatalog (): Promise<void> {
    await this.FTPClient.ensureDir(this.remoteCatalog)
  }

  @decoratorLogCall()
  private async openRemoteCatalog (): Promise<void> {
    await this.FTPClient.cd(this.remoteCatalog)
  }

  @decoratorLogCall()
  private async cleanRemoteCatalog (): Promise<void> {
    const remoteFiles = await this.FTPClient.list()

    for (const file of remoteFiles) {
      if (this.remoteIgnoreFile.includes(file.name)) {
        continue
      }

      const removeStrategies: RemoveStrategies = {
        0: async () => await Promise.resolve(),
        1: async name => await this.FTPClient.remove(name),
        2: async name => await this.FTPClient.removeDir(name),
        3: async () => await Promise.resolve()
      }
      await removeStrategies[file.type](file.name)
    }
  }

  @decoratorLogCall()
  private async uploadToRemoteCatalog (): Promise<void> {
    this.FTPClient.trackProgress(this.handlerUploadProgress)
    await this.FTPClient.uploadFromDir(this.localCatalog)
    this.FTPClient.trackProgress()
  }

  private handlerUploadProgress (info: ProgressInfo): void {
    if (info.bytes === 0) {
      process.stdout.write(`Uploading file ${info.name}\n`)
    }
  }

  @decoratorLogCall()
  private closeConnection (): void {
    this.FTPClient.close()
  }
}

export default DeployService
