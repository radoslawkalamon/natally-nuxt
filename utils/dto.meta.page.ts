export type DTOMetaPageConstructor = {
  createdAt: string;
  description: string;
  imageOpenGraph?: string;
  path: string;
  title: string;
  updatedAt: string;
}

const getPath = (path: string) => {
  const pathWithoutIndex = path.split('/').filter(i => i !== 'index').join('/')
  return `${pathWithoutIndex}${pathWithoutIndex.at(-1) !== '/' ? '/' : ''}`
}

export class DTOMetaPage {
  readonly createdAt: Date;
  readonly description: string;
  readonly imageOpenGraph: string;
  readonly path: string;
  readonly title: string;
  readonly updatedAt: Date;

  constructor ({
    createdAt,
    description,
    imageOpenGraph = 'common.opengraph.png',
    title,
    updatedAt,
    path
  }: DTOMetaPageConstructor) {
    this.createdAt = new Date(createdAt)
    this.description = description
    this.imageOpenGraph = `/images/${imageOpenGraph}`
    this.path = getPath(path)
    this.title = title
    this.updatedAt = new Date(updatedAt)
  }
}
