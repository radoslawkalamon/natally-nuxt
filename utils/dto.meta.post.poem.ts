import { DTOMetaPage, DTOMetaPageConstructor } from '@/utils/dto.meta.page'

export type DTOMetaPostPoemConstructor = DTOMetaPageConstructor & {
  imageCover: string
  imageOpenGraph: string
}

export class DTOMetaPostPoem extends DTOMetaPage {
  readonly imageCover: string;
  readonly imageOpenGraph: string;

  constructor ({
    createdAt,
    description,
    imageCover,
    imageOpenGraph,
    path,
    title,
    updatedAt
  }: DTOMetaPostPoemConstructor) {
    super({
      createdAt,
      description,
      path,
      title,
      updatedAt
    })

    this.imageCover = `${this.imagePath}${imageCover}`
    this.imageOpenGraph = `${this.imagePath}${imageOpenGraph}`
  }

  get imagePath (): string {
    return '/images/poezja-314/'
  }
}
