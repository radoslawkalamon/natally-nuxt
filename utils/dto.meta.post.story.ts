import { DTOMetaPage, DTOMetaPageConstructor } from '@/utils/dto.meta.page'

export type DTOMetaPostStoryConstructor = DTOMetaPageConstructor & {
  audiobookId?: string;
  imageCover: string;
  imageCover2x: string;
  imageOpenGraph: string
  timeReading: number;
}

export class DTOMetaPostStory extends DTOMetaPage {
  readonly audiobookId: string;
  readonly imageCover: string;
  readonly imageCover2x: string;
  readonly imageOpenGraph: string;
  readonly timeReading: number;

  constructor ({
    audiobookId = '0',
    createdAt,
    description,
    imageCover,
    imageCover2x,
    imageOpenGraph,
    path,
    timeReading,
    title,
    updatedAt
  }: DTOMetaPostStoryConstructor) {
    super({
      createdAt,
      description,
      path,
      title,
      updatedAt
    })

    this.audiobookId = audiobookId
    this.imageCover = `${this.imagePath}${imageCover}`
    this.imageCover2x = `${this.imagePath}${imageCover2x}`
    this.imageOpenGraph = `${this.imagePath}${imageOpenGraph}`
    this.timeReading = timeReading
  }

  get imagePath (): string {
    return '/images/'
  }
}
