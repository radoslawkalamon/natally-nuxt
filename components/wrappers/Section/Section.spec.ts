import Section from '@/components/wrappers/Section/Section.vue'
import { shallPassUnitSanityTest } from '@/devtools/jest.common.spec'
import { createDefaultOptionsFactory } from '@/devtools/jest.common.spec.utils'

const defaultOptionsFactory = createDefaultOptionsFactory({
  slots: {
    default: 'This is default slot test'
  }
})

describe('Wrappers / Section', () => {
  describe('Unit', () => {
    shallPassUnitSanityTest({
      component: Section,
      description: 'no footer slot',
      options: defaultOptionsFactory({
        slots: {
          header: 'This is header slot test'
        }
      })
    })

    shallPassUnitSanityTest({
      component: Section,
      description: 'no header slot',
      options: defaultOptionsFactory({
        slots: {
          footer: 'This is footer slot test'
        }
      })
    })
  })
})
