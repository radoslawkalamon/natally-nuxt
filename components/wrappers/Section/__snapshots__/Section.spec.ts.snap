// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wrappers / Section Unit shall pass unit sanity test | no footer slot 1`] = `
"<section class=\\"section\\">
  <header data-test=\\"wrappers-section-header\\" class=\\"section__header\\">This is header slot test</header>
  <div class=\\"section__inner\\">This is default slot test</div>
  <!---->
</section>"
`;

exports[`Wrappers / Section Unit shall pass unit sanity test | no header slot 1`] = `
"<section class=\\"section\\">
  <!---->
  <div class=\\"section__inner\\">This is default slot test</div>
  <footer data-test=\\"wrappers-section-footer\\" class=\\"section__footer\\">This is footer slot test</footer>
</section>"
`;
